cgen error, generics with implicit context

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

<details>
<summary>V version: V 0.4.9 1d9aa88, press to see full `v doctor` output</summary>

|V full version      |V 0.4.9 1d9aa88
|:-------------------|:-------------------
|OS                  |windows, Њ ©Єа®б®дв Windows 10 Pro 19045 64-а §ап¤­ п
|Processor           |28 cpus, 64bit, little endian, Intel(R) Xeon(R) CPU E5-2680 v4 @ 2.40GHz  
|Memory              |17.05GB/31.86GB
|                    |
|V executable        |C:\dev\v\v.exe
|V last modified time|2025-01-26 10:35:18
|                    |
|V home dir          |OK, value: C:\dev\v
|VMODULES            |OK, value: C:\Users\koplenov\.vmodules
|VTMP                |OK, value: C:\Users\koplenov\AppData\Local\Temp\v_0
|Current working dir |OK, value: C:\dev\mam_prod\v\litequeue\litequeue2
|                    |
|Git version         |git version 2.39.1.windows.1
|V git status        |Error: fatal: not a git repository (or any of the parent directories): .git
|.git/config present |false
|                    |
|cc version          |N/A
|gcc version         |N/A
|clang version       |N/A
|msvc version        |N/A
|tcc version         |tcc version 0.9.27 (x86_64 Windows)
|tcc git status      |N/A
|emcc version        |N/A
|glibc version       |N/A

</details>

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg  .\queue.v && C:\dev\mam_prod\v\litequeue\litequeue2\queue.exe`
See attached file `.\queue.v`

**What did you see?**
```
================== C compilation error (from tcc): ==============
cc: C:/Users/koplenov/AppData/Local/Temp/v_0/queue.01JQ7NSDSETKSY84VRGM1A7YM5.tmp.c:579: warning: WINVER redefined
cc: C:/Users/koplenov/AppData/Local/Temp/v_0/queue.01JQ7NSDSETKSY84VRGM1A7YM5.tmp.c:2096: error: 'Array_main__Message' undeclared
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

**What did you expect to see?**

Generics + Orm + Implicit context

Trying to make a message queue system on top of sqlite3, encountering a number of problems with generics

Version without generics and hidden contexts - https://github.com/koplenov/litequeue


Code:
``` vlang
module main

import time
import db.sqlite

pub enum MessageStatus as u8 {
	ready
	locked
	done
	failed
}

// Struct Message
pub struct Message[T] {
	data       string
	message_id int @[primary; sql: serial]
	status     MessageStatus
	in_time    time.Time @[default: 'CURRENT_TIME']
	lock_time  ?time.Time
	done_time  ?time.Time
	payload T
}

// Struct Queue
pub struct Queue[T] {
pub mut:
	conn 	&sqlite.DB
	handler ? fn (message Message[T]) !
}

fn C.sqlite3_config(int)

// Function NewQueue (Constructor)
pub struct Queue_config {
pub:
	path             string
	db               ?sqlite.DB      @[omitempty]
	pooling_interval i64 = time.second
}


pub fn new[T](config Queue_config) !Queue[T] {
	C.sqlite3_config(3)

	// use exist db connection or create file self
	mut conn := if db := config.db {
		db
	} else {
		sqlite.connect(config.path) or { return error('Failed to connect to database: ${err}') }
	}
	conn.busy_timeout(10 * 60_000)

	mut queue := Queue[T]{
		conn: &conn
	}

	sql queue.conn {
		create table Message[T]
	}!

	return queue
}



// Method setup
fn (mut self Queue[T]) process() {
		if handler := self.handler {
		println('taked handler function!')
		for {
			if message := self.take() {
				handler(message) or {
					self.mark_fail(message.message_id)
					println('[queue][fail][${message.message_id}]')
					continue
				}
				self.mark_done(message.message_id)
				println('[queue][done][${message.message_id}]')
			} else {
				// println('[queue][pooling]')
				time.sleep(500)
			}
		}
	} else {
		println('No handler!')
	}
}


// Method put
pub fn (mut self Queue[T]) add(payload T) !Message[T] {
	message := Message[T]{
		payload:    payload
		status:  MessageStatus.ready
		in_time: time.now()
	}

	sql self.conn {
		insert message into Message[T]
	} or { return err }
	return message
}

// Method pop_transaction
pub fn (mut self Queue[T]) take() ?Message[T] {
	now := time.now()

	messages := sql self.conn {
		select from Message[T] where status == MessageStatus.ready order by message_id limit 1
	} or { return none }

	if messages.len == 0 {
		return none
	}

	message := messages.first()

	sql self.conn {
		update Message[T] set status = MessageStatus.locked where message_id == message.message_id
		update Message[T] set lock_time = now where message_id == message.message_id
	} or { panic(err) }
	return message

}

// Return all the tasks in gived state.
pub fn (mut self Queue[T]) list_of(status MessageStatus) []Message {
	list_failed := sql self.conn {
		select from Message[T] where status == status
	} or { panic(err) }
	return list_failed
}

// Mark a locked message as free again.
pub fn (mut self Queue[T]) mark_fail(message_id int) {
	sql self.conn {
		update Message[T] set status = MessageStatus.failed where message_id == message_id
		update Message[T] set done_time = none where message_id == message_id
	} or { panic(err) }
}

// Method pop_transaction
pub fn (mut self Queue[T]) mark_done(message_id int) {
	now := time.now()

	sql self.conn {
		update Message[T] set status = MessageStatus.done where message_id == message_id
		update Message[T] set done_time = now where message_id == message_id
	} or { panic(err) }
}

// Mark a locked message as free again.
pub fn (mut self Queue[T]) len() int {
	count := sql self.conn {
		select count from Message[T]
	} or { panic(err) }
	return count
}

pub fn (mut self Queue[T]) next() ?Message[T] {
	return self.take()
}

// Return all the tasks in gived state.
pub fn (mut self Queue[T]) close() ! {
	self.conn.close()!
}



// usage
struct Payload {
	some_filed_1 string
	some_filed_2 int
}

fn test_handler(message Message[Payload]) ! {
	return error('AAAA')
}

fn main() {
	db_path := 'test_queue.db'
	mut queue := new[Payload](
		path: db_path
	) or { panic('Failed to create queue: ${err}') }

	queue.handler = test_handler
	spawn fn [mut queue]() {
		queue.process()
	}()
}



```
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/orm/orm_generic_test.v
+++ b/vlib/orm/orm_generic_test.v
@@ -1,3 +1,4 @@
+// vtest build: present_sqlite3? && !sanitize-memory-clang
 import db.sqlite
 
 pub enum MessageStatus as u8 {

--- a/vlib/orm/orm_generic_test.v
+++ b/vlib/orm/orm_generic_test.v
@@ -1,106 +1,47 @@
-module main
-
-import time
 import db.sqlite
 
 pub enum MessageStatus as u8 {
 	ready
-	locked
-	done
-	failed
 }
 
-// Struct Message
 pub struct Message[T] {
-	data       string
 	message_id int @[primary; sql: serial]
 	status     MessageStatus
-	in_time    time.Time @[default: 'CURRENT_TIME']
-	lock_time  ?time.Time
-	done_time  ?time.Time
-	payload    T
 }
 
-// Struct Queue
 pub struct Queue[T] {
-pub mut:
-	conn    &sqlite.DB
-	handler ?fn (message Message[T]) !
+	conn &sqlite.DB
 }
 
-fn C.sqlite3_config(int)
-
-// Function NewQueue (Constructor)
 pub struct Queue_config {
-pub:
-	path             string
-	db               ?sqlite.DB @[omitempty]
-	pooling_interval i64 = time.second
+	path string
+	db   ?sqlite.DB @[omitempty]
 }
 
 pub fn new[T](config Queue_config) !Queue[T] {
-	C.sqlite3_config(3)
-
-	// use exist db connection or create file self
 	mut conn := if db := config.db {
 		db
 	} else {
 		sqlite.connect(config.path) or { return error('Failed to connect to database: ${err}') }
 	}
-	conn.busy_timeout(10 * 60_000)
-
 	mut queue := Queue[T]{
 		conn: &conn
 	}
-
-	sql queue.conn {
-		create table Message[T]
-	}!
-
 	return queue
 }
 
-// Method put
-pub fn (mut self Queue[T]) add(payload T) !Message[T] {
-	message := Message[T]{
-		payload: payload
-		status:  MessageStatus.ready
-		in_time: time.now()
-	}
-	dump(message)
-
-	sql self.conn {
-		insert message into Message[T]
-	} or { return err }
+pub fn (mut self Queue[T]) take() !Message[T] {
+	messages := sql self.conn {
+		select from Message[T] where status == MessageStatus.ready order by message_id limit 1
+	} or { return error('') }
+	message := messages.first()
 	return message
 }
 
-// Method pop_transaction
-pub fn (mut self Queue[T]) take() ?[]Message[T] {
-	messages := sql self.conn {
-		select from Message[T] where status == MessageStatus.ready order by message_id limit 1
-	} or { return none }
-
-	if messages.len == 0 {
-		return none
-	}
-
-	return messages
-}
-
-// usage
-struct Payload {
-	some_filed_1 string
-	some_filed_2 int
-}
-
-fn test_handler(message Message[Payload]) ! {
-	return error('AAAA')
-}
+struct Payload {}
 
 fn test_main() {
-	db_path := ':memory:'
 	_ := new[Payload](
-		path: db_path
+		path: ':memory:'
 	) or { panic('Failed to create queue: ${err}') }
 }

--- a/vlib/orm/orm_generic_test.v
+++ b/vlib/orm/orm_generic_test.v
@@ -42,7 +42,5 @@ pub fn (mut self Queue[T]) take() !Message[T] {
 struct Payload {}
 
 fn test_main() {
-	_ := new[Payload](
-		path: ':memory:'
-	) or { panic('Failed to create queue: ${err}') }
+	_ := new[Payload](path: ':memory:')!
 }

--- a/vlib/orm/orm_generic_test.v
+++ b/vlib/orm/orm_generic_test.v
@@ -1,0 +1,106 @@
+module main
+
+import time
+import db.sqlite
+
+pub enum MessageStatus as u8 {
+	ready
+	locked
+	done
+	failed
+}
+
+// Struct Message
+pub struct Message[T] {
+	data       string
+	message_id int @[primary; sql: serial]
+	status     MessageStatus
+	in_time    time.Time @[default: 'CURRENT_TIME']
+	lock_time  ?time.Time
+	done_time  ?time.Time
+	payload    T
+}
+
+// Struct Queue
+pub struct Queue[T] {
+pub mut:
+	conn    &sqlite.DB
+	handler ?fn (message Message[T]) !
+}
+
+fn C.sqlite3_config(int)
+
+// Function NewQueue (Constructor)
+pub struct Queue_config {
+pub:
+	path             string
+	db               ?sqlite.DB @[omitempty]
+	pooling_interval i64 = time.second
+}
+
+pub fn new[T](config Queue_config) !Queue[T] {
+	C.sqlite3_config(3)
+
+	// use exist db connection or create file self
+	mut conn := if db := config.db {
+		db
+	} else {
+		sqlite.connect(config.path) or { return error('Failed to connect to database: ${err}') }
+	}
+	conn.busy_timeout(10 * 60_000)
+
+	mut queue := Queue[T]{
+		conn: &conn
+	}
+
+	sql queue.conn {
+		create table Message[T]
+	}!
+
+	return queue
+}
+
+// Method put
+pub fn (mut self Queue[T]) add(payload T) !Message[T] {
+	message := Message[T]{
+		payload: payload
+		status:  MessageStatus.ready
+		in_time: time.now()
+	}
+	dump(message)
+
+	sql self.conn {
+		insert message into Message[T]
+	} or { return err }
+	return message
+}
+
+// Method pop_transaction
+pub fn (mut self Queue[T]) take() ?[]Message[T] {
+	messages := sql self.conn {
+		select from Message[T] where status == MessageStatus.ready order by message_id limit 1
+	} or { return none }
+
+	if messages.len == 0 {
+		return none
+	}
+
+	return messages
+}
+
+// usage
+struct Payload {
+	some_filed_1 string
+	some_filed_2 int
+}
+
+fn test_handler(message Message[Payload]) ! {
+	return error('AAAA')
+}
+
+fn test_main() {
+	db_path := ':memory:'
+	_ := new[Payload](
+		path: db_path
+	) or { panic('Failed to create queue: ${err}') }
+}

--- a/vlib/v/parser/orm.v
+++ b/vlib/v/parser/orm.v
@@ -101,6 +101,8 @@ fn (mut p Parser) sql_expr() ast.Expr {
 
 	if is_count {
 		typ = ast.int_type
+	} else if table_type.has_flag(.generic) {
+		typ = ast.new_type(p.table.find_or_register_array(table_type)).set_flag(.generic)
 	} else {
 		typ = ast.new_type(p.table.find_or_register_array(table_type))
 	}

